File: generated_code/optimized/runtime/python/1.py
from typing import List


def separate_paren_groups(paren_string: str) -> List[str]:
    s = ''.join(c for c in paren_string if c == '(' or c == ')')
    result = []
    depth = 0
    start = 0

    for i, c in enumerate(s):
        if c == '(':
            depth += 1
        else:
            depth -= 1
            if depth == 0:
                result.append(s[start:i + 1])
                start = i + 1

    return result

File: generated_code/optimized/balanced/detailed/cpp/1.cpp
#include <vector>
#include <string>

using namespace std;

vector<string> separate_paren_groups(string paren_string) {
    vector<string> all_parens;
    int level = 0;
    size_t start = 0;
    const size_t n = paren_string.size();

    for (size_t i = 0; i < n; ++i) {
        const char c = paren_string[i];
        if (c == '(') {
            if (level == 0) start = i;
            ++level;
        } else if (c == ')') {
            --level;
            if (level == 0) {
                all_parens.emplace_back(paren_string, start, i - start + 1);
            }
        }
    }

    return all_parens;
}
